maitake/src/wait/cell.rs

```diff
@@ -56,26 +91,24 @@ impl WaitCell {
 }
 
 impl WaitCell {
-    pub fn poll_wait(&self, waker: &Waker) -> Poll<WaitResult<()>> {
+    fn register_wait(&self, waker: &Waker) -> Result<(), Error> {
         tracing::trace!(wait_cell = ?fmt::ptr(self), ?waker, "registering waker");
 
         // this is based on tokio's AtomicWaker synchronization strategy
         match test_dbg!(self.compare_exchange(State::WAITING, State::PARKING, Acquire)) {
             // someone else is notifying, so don't wait!
             Err(actual) if test_dbg!(actual.is(State::CLOSED)) => {
-                return wait::closed();
+                return closed();
             }
             Err(actual) if test_dbg!(actual.is(State::NOTIFYING)) => {
-                waker.wake_by_ref();
-                crate::loom::hint::spin_loop();
-                return wait::notified(());
+                return busy();
             }
 
             Err(actual) => {
                 debug_assert!(
                     actual == State::PARKING || actual == State::PARKING | State::NOTIFYING
                 );
-                return Poll::Pending;
+                return busy();
```

Comment:
If we're in the middle of some other operation, we can't possibly have added this new waker to the cell. Report it as busy.